pkg/image/trigger/buildconfigs/buildconfigs.go

```diff
@@ -0,0 +1,194 @@
+		if t.ImageChange.From != nil {
+			from = t.ImageChange.From
+			fieldPath = "spec.triggers"
+		} else {
+			from = buildutil.GetInputReference(bc.Spec.Strategy)
+			fieldPath = "spec.strategy.*.from"
+		}
+		if from == nil || from.Kind != "ImageStreamTag" || len(from.Name) == 0 {
+			continue
+		}
+
+		// add a trigger
+		triggers = append(triggers, triggerapi.ObjectFieldTrigger{
+			From: triggerapi.ObjectReference{
+				Name:       from.Name,
+				Namespace:  from.Namespace,
+				Kind:       from.Kind,
+				APIVersion: from.APIVersion,
+			},
+			FieldPath: fieldPath,
+		})
+	}
+	return triggers
+}
+
+// buildConfigTriggerIndexer converts build config events into entries for the trigger cache, and
+// also calculates the latest state of the changes on the object.
+type buildConfigTriggerIndexer struct {
+	prefix string
+}
+
+func NewBuildConfigTriggerIndexer(prefix string) trigger.Indexer {
+	return buildConfigTriggerIndexer{prefix: prefix}
+}
+
+func (i buildConfigTriggerIndexer) Index(obj, old interface{}) (string, *trigger.CacheEntry, cache.DeltaType, error) {
+	var (
+		triggers []triggerapi.ObjectFieldTrigger
+		bc       *buildapi.BuildConfig
+		change   cache.DeltaType
+	)
+	switch {
+	case obj != nil && old == nil:
+		// added
+		bc = obj.(*buildapi.BuildConfig)
+		triggers = calculateBuildConfigTriggers(bc)
+		change = cache.Added
+	case old != nil && obj == nil:
+		// deleted
+		bc = old.(*buildapi.BuildConfig)
+		triggers = calculateBuildConfigTriggers(bc)
+		change = cache.Deleted
+	default:
+		// updated
+		bc = obj.(*buildapi.BuildConfig)
+		triggers = calculateBuildConfigTriggers(bc)
+		oldTriggers := calculateBuildConfigTriggers(old.(*buildapi.BuildConfig))
+		switch {
+		case len(oldTriggers) == 0:
+			change = cache.Added
+		case !reflect.DeepEqual(oldTriggers, triggers):
+			change = cache.Updated
+		}
+	}
+
+	if len(triggers) > 0 {
+		key := i.prefix + bc.Namespace + "/" + bc.Name
+		return key, &trigger.CacheEntry{
+			Key:       key,
+			Namespace: bc.Namespace,
+			Triggers:  triggers,
+		}, change, nil
+	}
+	return "", nil, change, nil
+}
+
+// BuildConfigInstantiator abstracts creating builds from build requests.
+type BuildConfigInstantiator interface {
+	// Instantiate should launch a build from the provided build request.
+	Instantiate(namespace string, request *buildapi.BuildRequest) (*buildapi.Build, error)
+}
+
+// BuildConfigReactor converts trigger changes into new builds. It will request a build if
+// at least one image is out of date.
+type BuildConfigReactor struct {
+	Instantiator BuildConfigInstantiator
+}
+
+// ImageChanged is passed a build config and a set of changes and updates the object if
+// necessary.
+func (r *BuildConfigReactor) ImageChanged(obj interface{}, tagRetriever trigger.TagRetriever) error {
+	bc := obj.(*buildapi.BuildConfig)
+
+	var (
+		changed bool
+		id      string
+		from    *kapi.ObjectReference
+		ref     string
+	)
+	for _, t := range bc.Spec.Triggers {
+		p := t.ImageChange
```

Comment:
P stands for "Params" out of "imageChangeParams" and is used consistently as a short local variable in these triggers.